src/ucs/sys/sock.c

```diff
@@ -582,28 +582,49 @@ ucs_socket_handle_io(int fd, const void *data, size_t count,
 
 static inline ucs_status_t
 ucs_socket_do_io_nb(int fd, void *data, size_t *length_p,
-                    ucs_socket_io_func_t io_func, const char *name)
+                    ucs_socket_io_func_t io_func, const char *name, int flags)
 {
-    ssize_t ret = io_func(fd, data, *length_p, MSG_NOSIGNAL);
+    ssize_t ret;
+
+    if (flags == 0) {
+        flags = MSG_NOSIGNAL;
+    }
+
+    ret = io_func(fd, data, *length_p, flags);
     return ucs_socket_handle_io(fd, data, *length_p, length_p, 0,
                                 ret, errno, name);
 }
 
 static inline ucs_status_t
-ucs_socket_do_io_b(int fd, void *data, size_t length,
-                   ucs_socket_io_func_t io_func, const char *name)
+ucs_socket_do_io_b(int fd, void *data, size_t *length,
+                   ucs_socket_io_func_t io_func, const char *name, int flags)
 {
-    size_t done_cnt = 0, cur_cnt = length;
+    size_t done_cnt = 0, cur_cnt = *length;
     ucs_status_t status;
+    int type;
+
+    status = ucs_socket_getopt(fd, SOL_SOCKET, SO_TYPE, (void*)&type,
+                               sizeof(type));
+    if (status != UCS_OK) {
+        return UCS_ERR_IO_ERROR;
+    }
 
     do {
-        status = ucs_socket_do_io_nb(fd, data, &cur_cnt, io_func, name);
+        status = ucs_socket_do_io_nb(fd, data, &cur_cnt, io_func, name, flags);
         done_cnt += cur_cnt;
-        ucs_assert(done_cnt <= length);
-        cur_cnt = length - done_cnt;
-    } while ((done_cnt < length) &&
+        if ((flags & MSG_TRUNC) == 0) {
+            ucs_assert(done_cnt <= *length);
+        }
```

Comment:
But I guess as long as it supports receiving flags we should handle MSG_TRUNC, otherwise it crashes for `done_cnt <= *length`.
Another option is to not support flags at all